### PR TITLE
[#67] Agent

### PR DESCRIPTION
**Pull request by JanKapala**
Implemented:

- Services History Generator (Closes #80)
- Base Agent Recommender (Closes #78) 
- RL Agent Recommender (Closes #81)

### CODE REVIEW

File: recommender/engine/rl_agent/rl_agent.py [hunk 1]
```diff
@@ -48,19 +77,26 @@ def _for_logged_user(self, user: User, search_data: SearchData, k: int) -> Tuple
         state_tensors = self.state_embedder(state)
 
         weights_tensor = actor_model(*state_tensors)
-        recommended_services = self.action_selector(weights_tensor, user, search_data) # TODO: jednak ids
-        recommended_services_ids = _services_to_ids(recommended_services)
+        recommended_services_ids = self.action_selector(
```
Comment (wujuu) [inline recommender/engine/rl_agent/rl_agent.py:80]: "recommended_services_ids" -> "recommended_service_ids". (Same as rails defines methods for plural resources as this -> `User.services` and `User.service_ids`)

Comment (wujuu) [reply recommender/engine/rl_agent/rl_agent.py:80]: This applies to every variable and method named this way, such as `get_accessed_services_ids`. This is a small change, but looks more proffesional if someone reads our code

File: recommender/engine/rl_agent/rl_agent.py [hunk 2]
```diff
@@ -70,5 +106,14 @@ def _create_state(self, user: User, search_data: SearchData) -> State:
             state: State containing information about user and search_data
         """
 
-        # TODO: implement
-        pass
+        search_data = copy.deepcopy(search_data)
+        search_data.pop("rating")
```
Comment (wujuu) [inline recommender/engine/rl_agent/rl_agent.py:110]: I would just add "rating" to SearchData model - looks better and we can always just ignore it later if needed

File: recommender/errors.py [hunk 3]
```diff
@@ -11,7 +11,12 @@ def message(self):  # pragma: no cover
         return "Invalid recommendation panel id error"
 
 
-class UntrainedPreAgentError(Exception):
+class NoActorModelForK(RecommendationEngineError):
+    def message(self):  # pragma: no cover
+        return "There is no Actor model for used K (number of services in recommendation)."
```
Comment (wujuu) [inline recommender/errors.py:16]: "for used K" -> "for K used"

File: recommender/errors.py [hunk 4]
```diff
@@ -11,7 +11,12 @@ def message(self):  # pragma: no cover
...
+
+
+class UntrainedPreAgentError(RecommendationEngineError):
     def message(self):  # pragma: no cover
         return (
             "Pre-Agent can't operate without trained Neural Collaborative"
```
Comment (wujuu) [inline recommender/errors.py:22]: "without trained" -> "without a trained"

File: recommender/errors.py [hunk 5]
```diff
@@ -20,6 +25,11 @@ def message(self):  # pragma: no cover
         )
 
 
+class UntrainedRLAgentError(RecommendationEngineError):
+    def message(self):  # pragma: no cover
+        return "RL-Agent can't operate without trained actor models"
```
Comment (wujuu) [inline recommender/errors.py:30]: same as above

**Review by wujuu — CHANGES_REQUESTED** [review]
Generally ok, some small errors

File: recommender/engine/rl_agent/rl_agent.py (hunk 2 above)
Comment (JanKapala) [reply recommender/engine/rl_agent/rl_agent.py:110]: There was a discussion about adding rating to the SearchData model and the result was: not to add. So we will stay with it.

File: recommender/errors.py (hunk 3 above)
Comment (JanKapala) [reply recommender/errors.py:16]: "used" word is related to the "K" - not to the "Actor Model" so I think the grammar of this sentence is correct.

File: recommender/errors.py (hunk 5 above)
Comment (JanKapala) [reply recommender/errors.py:30]: Now, this is plural (model**s**), so no "a" before "trained"

**Review by wujuu — APPROVED** [review]

File: recommender/engine/base_agent_recommender.py [hunk 6]
```diff
@@ -38,15 +45,13 @@ def call(self, context: Dict[str, Any]) -> Tuple[int]:
 
         if user is not None:
             return self._for_logged_user(user, search_data, K)
-        else:
-            return self._for_anonymous_user(search_data, K)
+        return self._for_anonymous_user(search_data, K)
 
     def _load_models(self) -> None:
```
Comment (wujuu) [inline recommender/engine/base_agent_recommender.py:50]: annotate this class as abstract

File: recommender/engine/base_agent_recommender.py [hunk 7]
```diff
@@ -81,31 +86,61 @@ def _get_user(self, context: Dict[str, Any]) -> User:
 
         return user
 
-    def _for_logged_user(self, user: User, search_data: SearchData, k: int) -> Tuple[int]:
+    def _for_logged_user(
```
Comment (wujuu) [inline recommender/engine/base_agent_recommender.py:89]: annotate this class as abstract

File: recommender/engine/rl_agent/rl_agent.py [hunk 8]
```diff
@@ -1,45 +1,74 @@
...
+        actor_model_3=None,
+        state_embedder=None,
+        action_selector=None,
+    ) -> None:
         super().__init__()
-        self.actor_model_2, self.actor_model_3 = actor_model_2, actor_model_3
-        self.state_embedder = StateEmbedder()
-        self.action_selector = ActionSelector()
-        # TODO: loading from database if None
+        self.actor_model_2 = actor_model_2
```
Comment (wujuu) [inline recommender/engine/rl_agent/rl_agent.py:32]: please name these variables as actor_model_v1 and actor_model_v2

File: recommender/engine/rl_agent/rl_agent.py [hunk 9]
```diff
@@ -1,45 +1,74 @@
...
 
-        # TODO: implement lazy models (and maybe other things) loading
-        self.actor_model_2, self.actor_model_3 = None, None
-        pass
-
-    def _for_logged_user(self, user: User, search_data: SearchData, k: int) -> Tuple[int]:
+        if self.actor_model_2 is None:
+            try:
+                self.actor_model_2 = load_last_module(
+                    ACTOR_MODEL_V2
```
Comment (wujuu) [inline recommender/engine/rl_agent/rl_agent.py:46]: ACTOR_MODEL_V1

File: recommender/engine/rl_agent/rl_agent.py [hunk 10]
```diff
@@ -1,45 +1,74 @@
...
+                )  # TODO: import this constant from actor_model.py
+            except NoSavedModuleError as no_saved_module:
+                raise UntrainedRLAgentError from no_saved_module
+
+        if self.actor_model_3 is None:
+            try:
+                self.actor_model_3 = load_last_module(
+                    ACTOR_MODEL_V3
```
Comment (wujuu) [inline recommender/engine/rl_agent/rl_agent.py:54]: ACTOR_MODEL_V2

**Review by wujuu — CHANGES_REQUESTED** [review]

File: recommender/engine/base_agent_recommender.py (hunk 6 above)
Comment (JanKapala) [reply recommender/engine/base_agent_recommender.py:50]: this is not abstract method (nor class xD)

File: recommender/engine/base_agent_recommender.py (hunk 7 above)
Comment (JanKapala) [reply recommender/engine/base_agent_recommender.py:89]: okey, I will annotate this abstract **method** :)

**Review by wujuu — APPROVED** [review]

**Pull request merged**